Index: lib/net/connection.cpp
===================================================================
--- lib/net/connection.cpp	(revision 3455)
+++ lib/net/connection.cpp	(working copy)
@@ -120,7 +120,8 @@
     readNB( buffer, bytes );
 }
 
-bool Connection::recvSync( void** outBuffer, uint64_t* outBytes )
+bool Connection::recvSync( void** outBuffer, uint64_t* outBytes,
+                           const bool wait )
 {
     EQASSERT( _aioBuffer );
     EQASSERT( _aioBytes );
@@ -130,7 +131,7 @@
     if( outBytes )
         *outBytes = _aioBytes;
 
-    void* buffer( _aioBuffer );
+    void* const buffer( _aioBuffer );
     const uint64_t bytes( _aioBytes );
     _aioBuffer = 0;
     _aioBytes  = 0;
@@ -143,7 +144,7 @@
 
     while( bytesLeft )
     {
-        const int64_t got = readSync( ptr, bytesLeft );
+        const int64_t got = readSync( ptr, bytesLeft, wait );
 
         if( got < 0 ) // error
         {
@@ -164,7 +165,10 @@
             {
                 if( outBytes )
                     *outBytes = 0;
-                return false;
+
+                _aioBuffer = buffer; // Operation still pending, restore data
+                _aioBytes = bytes;
+                return true;
             }
             EQVERB << "Zero bytes read" << endl;
         }
Index: lib/net/node.cpp
===================================================================
--- lib/net/node.cpp	(revision 3455)
+++ lib/net/node.cpp	(working copy)
@@ -830,13 +830,15 @@
 
     void* sizePtr( 0 );
     uint64_t bytes( 0 );
-    const bool gotSize = connection->recvSync( &sizePtr, &bytes );
-
-    if( !gotSize ) // Some systems signal data on dead connections.
+    const bool readOK = connection->recvSync( &sizePtr, &bytes, false );
+  
+    if( !readOK ) // Some systems signal data on dead connections.
     {
         connection->recvNB( sizePtr, sizeof( uint64_t ));
         return false;
     }
+    if( bytes == 0 ) // ConnectionSet signalled data but there is none.
+        return true; // IO still pending, don't start a new one
 
     EQASSERT( sizePtr );
     const uint64_t size = *reinterpret_cast< uint64_t* >( sizePtr );
Index: lib/net/socketConnection.h
===================================================================
--- lib/net/socketConnection.h	(revision 3455)
+++ lib/net/socketConnection.h	(working copy)
@@ -67,7 +67,8 @@
 
 #ifdef WIN32
         virtual void readNB( void* buffer, const uint64_t bytes );
-        virtual int64_t readSync( void* buffer, const uint64_t bytes );
+        virtual int64_t readSync( void* buffer, const uint64_t bytes,
+                                  const bool wait );
         virtual int64_t write( const void* buffer, const uint64_t bytes );
 
         typedef SOCKET Socket;
Index: lib/net/pipeConnection.h
===================================================================
--- lib/net/pipeConnection.h	(revision 3455)
+++ lib/net/pipeConnection.h	(working copy)
@@ -62,7 +62,8 @@
     protected:
 #ifdef WIN32
         virtual void readNB( void* buffer, const uint64_t bytes );
-        virtual int64_t readSync( void* buffer, const uint64_t bytes );
+        virtual int64_t readSync( void* buffer, const uint64_t bytes,
+                                  const bool wait );
         virtual int64_t write( const void* buffer, const uint64_t bytes );
 #endif
 
Index: lib/net/fdConnection.h
===================================================================
--- lib/net/fdConnection.h	(revision 3455)
+++ lib/net/fdConnection.h	(working copy)
@@ -42,7 +42,8 @@
         FDConnection();
 
         virtual void readNB( void* buffer, const uint64_t bytes );
-        virtual int64_t readSync( void* buffer, const uint64_t bytes );
+        virtual int64_t readSync( void* buffer, const uint64_t bytes,
+                                  const bool wait );
         virtual int64_t write( const void* buffer, const uint64_t bytes );
 
         int   _readFD;     //!< The read file descriptor.
Index: lib/net/pairConnection.h
===================================================================
--- lib/net/pairConnection.h	(revision 3455)
+++ lib/net/pairConnection.h	(working copy)
@@ -47,8 +47,9 @@
 
         virtual void readNB( void* buffer, const uint64_t bytes )
             { _readConnection->readNB( buffer, bytes ); }
-        virtual int64_t readSync( void* buffer, const uint64_t bytes )
-            { return _readConnection->readSync( buffer, bytes ); }
+        virtual int64_t readSync( void* buffer, const uint64_t bytes,
+                                  const bool wait )
+            { return _readConnection->readSync( buffer, bytes, wait ); }
         virtual int64_t write( const void* buffer, const uint64_t bytes )
             { return _writeConnection->write( buffer, bytes ); }
 
Index: lib/net/namedPipeConnection.h
===================================================================
--- lib/net/namedPipeConnection.h	(revision 3455)
+++ lib/net/namedPipeConnection.h	(working copy)
@@ -70,7 +70,8 @@
 
 #ifdef WIN32
         virtual void readNB( void* buffer, const uint64_t bytes );
-        virtual int64_t readSync( void* buffer, const uint64_t bytes );
+        virtual int64_t readSync( void* buffer, const uint64_t bytes,
+                                  const bool wait );
         virtual int64_t write( const void* buffer, const uint64_t bytes );
 #else
 
Index: lib/net/bufferConnection.h
===================================================================
--- lib/net/bufferConnection.h	(revision 3455)
+++ lib/net/bufferConnection.h	(working copy)
@@ -41,7 +41,8 @@
     protected:
         virtual void readNB( void* buffer, const uint64_t bytes )
             { EQDONTCALL; }
-        virtual int64_t readSync( void* buffer, const uint64_t bytes )
+        virtual int64_t readSync( void* buffer, const uint64_t bytes,
+                                  const bool wait )
             { EQDONTCALL; return -1; }
         EQ_EXPORT virtual int64_t write( const void* buffer,
                                          const uint64_t bytes );
Index: lib/net/pipeConnectionWin32.cpp
===================================================================
--- lib/net/pipeConnectionWin32.cpp	(revision 3455)
+++ lib/net/pipeConnectionWin32.cpp	(working copy)
@@ -102,7 +102,9 @@
     _fireStateChanged();
 }
 void PipeConnection::readNB( void* buffer, const uint64_t bytes ) { /* NOP */ }
-int64_t PipeConnection::readSync( void* buffer, const uint64_t bytes )
+
+int64_t PipeConnection::readSync( void* buffer, const uint64_t bytes,
+                                  const bool wait )
 {
     if( !_readHandle )
         return -1;
Index: lib/net/connection.h
===================================================================
--- lib/net/connection.h	(revision 3455)
+++ lib/net/connection.h	(working copy)
@@ -200,11 +200,15 @@
          * and bytes return value pointers can be 0. This method uses readNB()
          * and readSync() to fill a buffer, potentially by using multiple reads.
          *
+         * A return value of true does not mean that all data has been read. The
+         * number of bytes read is returned in bytes.
+         *
          * @param buffer return value, the buffer pointer passed to recvNB().
          * @param bytes return value, the number of bytes read.
-         * @return true if all requested data has been read, false otherwise.
+         * @return false on error, true otherwise.
          */
-        EQ_EXPORT bool recvSync( void** buffer, uint64_t* bytes );
+        EQ_EXPORT bool recvSync( void** buffer, uint64_t* bytes, 
+                                 const bool wait = true );
 
         void getRecvData( void** buffer, uint64_t* bytes )
             { *buffer = _aioBuffer; *bytes = _aioBytes; }
@@ -234,7 +238,8 @@
          * @param bytes the number of bytes to read.
          * @return the number of bytes read, or -1 upon error.
          */
-        virtual int64_t readSync( void* buffer, const uint64_t bytes ) = 0;
+        virtual int64_t readSync( void* buffer, const uint64_t bytes,
+                                  const bool wait ) = 0;
         //@}
 
         /** @name Synchronous write to the connection */
Index: lib/net/namedPipeConnectionWin32.cpp
===================================================================
--- lib/net/namedPipeConnectionWin32.cpp	(revision 3455)
+++ lib/net/namedPipeConnectionWin32.cpp	(working copy)
@@ -326,7 +326,8 @@
     }
 }
 
-int64_t NamedPipeConnection::readSync( void* buffer, const uint64_t bytes )
+int64_t NamedPipeConnection::readSync( void* buffer, const uint64_t bytes,
+                                       const bool wait )
 {
     CHECK_THREAD( _recvThread );
 
Index: lib/net/fdConnection.cpp
===================================================================
--- lib/net/fdConnection.cpp	(revision 3455)
+++ lib/net/fdConnection.cpp	(working copy)
@@ -43,7 +43,8 @@
 //----------------------------------------------------------------------
 void FDConnection::readNB( void* buffer, const uint64_t bytes ) { /* NOP */ }
 
-int64_t FDConnection::readSync( void* buffer, const uint64_t bytes )
+int64_t FDConnection::readSync( void* buffer, const uint64_t bytes,
+                                const bool wait )
 {
     if( _readFD < 1 )
         return -1;
Index: lib/net/socketConnectionWin32.cpp
===================================================================
--- lib/net/socketConnectionWin32.cpp	(revision 3455)
+++ lib/net/socketConnectionWin32.cpp	(working copy)
@@ -267,7 +267,8 @@
     }
 }
 
-int64_t SocketConnection::readSync( void* buffer, const uint64_t bytes )
+int64_t SocketConnection::readSync( void* buffer, const uint64_t bytes,
+                                    const bool wait )
 {
     CHECK_THREAD( _recvThread );
 
@@ -277,28 +278,34 @@
         return -1;
     }
 
-    DWORD got   = 0;
-    DWORD flags = 0;
-    if( !WSAGetOverlappedResult( _readFD, &_overlapped, &got, TRUE, &flags ))
+    while( true )
     {
-        if( GetLastError() == WSASYSCALLFAILURE ) // happens sometimes!?
-            return 0;
+        DWORD got   = 0;
+        DWORD flags = 0;
+        if( !WSAGetOverlappedResult( _readFD, &_overlapped, &got, wait, &flags ))
+        {
+            if( GetLastError() == WSASYSCALLFAILURE ) // happens sometimes!?
+                continue;
 
-        EQWARN << "Read complete failed: " << EQ_SOCKET_ERROR 
-               << ", closing connection" << std::endl;
-        close();
-        return -1;
-    }
+            if( GetLastError() == WSA_IO_INCOMPLETE )
+                return 0;
 
-    if( got == 0 )
-    {
-        EQWARN << "Read operation returned with nothing read"
-               << ", closing connection." << std::endl;
-        close();
-        return -1;
-    }
+            EQWARN << "Read complete failed: " << EQ_SOCKET_ERROR 
+                   << ", closing connection" << std::endl;
+            close();
+            return -1;
+        }
+
+        if( got == 0 )
+        {
+            EQWARN << "Read operation returned with nothing read"
+                   << ", closing connection." << std::endl;
+            close();
+            return -1;
+        }
 
-    return got;
+        return got;
+    }
 }
 
 int64_t SocketConnection::write( const void* buffer, const uint64_t bytes)
Index: lib/client/X11Connection.h
===================================================================
--- lib/client/X11Connection.h	(revision 3455)
+++ lib/client/X11Connection.h	(working copy)
@@ -58,7 +58,8 @@
     protected:
         virtual void readNB( void* buffer, const uint64_t bytes )
             { EQDONTCALL; }
-        virtual int64_t readSync( void* buffer, const uint64_t bytes )
+        virtual int64_t readSync( void* buffer, const uint64_t bytes,
+                                  const bool wait )
             { EQDONTCALL; return -1; }
         virtual int64_t write( const void* buffer, const uint64_t bytes )
             { EQDONTCALL; return -1; }
